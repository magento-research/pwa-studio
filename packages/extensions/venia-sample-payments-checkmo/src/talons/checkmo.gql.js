import { gql } from '@apollo/client';

export const GET_CHECKMO_CONFIG_DATA = gql`
    query storeConfigData {
        storeConfig {
            id
            payment_checkmo_payable_to @client
            payment_checkmo_mailing_address @client
        }
    }
`;

export const GET_IS_BILLING_ADDRESS_SAME = gql`
    query getIsBillingAddressSame($cartId: String!) {
        cart(cart_id: $cartId) @client {
            id
            isBillingAddressSame
        }
    }
`;

export const GET_BILLING_ADDRESS = gql`
    query getBillingAddress($cartId: String!) {
        cart(cart_id: $cartId) {
            id
            billingAddress: billing_address {
                firstName: firstname
                lastName: lastname
                country {
                    code
                }
                street
                city
                region {
                    code
                    label
                    region_id
                }
                postcode
                phoneNumber: telephone
            }
        }
    }
`;

export const GET_SHIPPING_ADDRESS = gql`
    query getSelectedShippingAddress($cartId: String!) {
        cart(cart_id: $cartId) {
            id
            shippingAddresses: shipping_addresses {
                firstName: firstname
                lastName: lastname
                country {
                    code
                }
                street
                city
                region {
                    code
                    region_id
                }
                postcode
                phoneNumber: telephone
            }
        }
    }
`;

export const SET_BILLING_ADDRESS = gql`
    mutation setBillingAddress(
        $cartId: String!
        $firstName: String!
        $lastName: String!
        $street1: String!
        $street2: String
        $city: String!
        $region: String!
        $postcode: String!
        $country: String!
        $phoneNumber: String!
    ) {
        setBillingAddressOnCart(
            input: {
                cart_id: $cartId
                billing_address: {
                    address: {
                        firstname: $firstName
                        lastname: $lastName
                        street: [$street1, $street2]
                        city: $city
                        region: $region
                        postcode: $postcode
                        country_code: $country
                        telephone: $phoneNumber
                        save_in_address_book: false
                    }
                }
            }
        ) @connection(key: "setBillingAddressOnCart") {
            cart {
                id
                billing_address {
                    firstname
                    lastname
                    country {
                        code
                    }
                    street
                    city
                    region {
                        code
                        label
                        region_id
                    }
                    postcode
                    telephone
                }
            }
        }
    }
`;

export const SET_PAYMENT_METHOD_ON_CART = gql`
    mutation setPaymentMethodOnCart($cartId: String!) {
        setPaymentMethodOnCart(
            input: { cart_id: $cartId, payment_method: { code: "checkmo" } }
        ) @connection(key: "setPaymentMethodOnCart") {
            cart {
                id
                selected_payment_method {
                    code
                    title
                }
            }
        }
    }
`;

export default {
    getCheckmoConfigQuery: GET_CHECKMO_CONFIG_DATA,
    getBillingAddressQuery: GET_BILLING_ADDRESS,
    getIsBillingAddressSameQuery: GET_IS_BILLING_ADDRESS_SAME,
    getShippingAddressQuery: GET_SHIPPING_ADDRESS,
    setBillingAddressMutation: SET_BILLING_ADDRESS,
    setPaymentMethodOnCartMutation: SET_PAYMENT_METHOD_ON_CART
};
