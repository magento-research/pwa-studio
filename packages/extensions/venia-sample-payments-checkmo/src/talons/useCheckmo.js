import { useCallback, useEffect, useMemo } from 'react';
import { useApolloClient, useQuery, useMutation } from '@apollo/client';
import { useFormState, useFormApi } from 'informed';

import mergeOperations from '@magento/peregrine/lib/util/shallowMerge';
import { useCartContext } from '@magento/peregrine/lib/context/cart';

import DEFAULT_OPERATIONS from './checkmo.gql';

/**
 * Maps address response data from GET_BILLING_ADDRESS and GET_SHIPPING_ADDRESS
 * queries to input names in the billing address form.
 * {@link checkmo.gql.js}.
 *
 * @param {ShippingCartAddress|BillingCartAddress} rawAddressData query data
 */
export const mapAddressData = rawAddressData => {
    if (!rawAddressData) return {};

    const {
        firstName,
        lastName,
        city,
        postcode,
        phoneNumber,
        street,
        country,
        region
    } = rawAddressData;

    return {
        firstName,
        lastName,
        city,
        postcode,
        phoneNumber,
        street1: street[0],
        street2: street[1],
        country: country.code,
        region: region.region_id
    };
};

/**
 * Talon to handle checkmo payment.
 *
 * @param {Boolean} props.shouldSubmit boolean value which represents if a payment nonce request has been submitted
 * @param {Function} props.onPaymentSuccess callback to invoke when the a payment nonce has been generated
 * @param {Function} props.onPaymentError callback to invoke when component throws an error
 * @param {Function} props.resetShouldSubmit callback to reset the shouldSubmit flag
 * @param {DocumentNode} props.operations.getCheckmoConfigQuery query to fetch config from backend
 * @param {DocumentNode} props.operations.setPaymentMethodOnCartMutation mutation to set checkmo as payment
 *
 * @returns {
 *  payableTo: String,
 *  mailingAddress: String,
 *  onBillingAddressChangedError: Function,
 *  onBillingAddressChangedSuccess: Function
 * }
 */
export const useCheckmo = props => {
    const {
        mailingAddress: propsMailingAddress,
        resetShouldSubmit,
        onPaymentSuccess,
        onPaymentError,
        shouldSubmit
    } = props;

    const operations = mergeOperations(DEFAULT_OPERATIONS, props.operations);

    const {
        getIsBillingAddressSameQuery,
        getCheckmoConfigQuery,
        getBillingAddressQuery,
        getShippingAddressQuery,
        setBillingAddressMutation,
        setPaymentMethodOnCartMutation
    } = operations;

    const client = useApolloClient();
    const [{ cartId }] = useCartContext();
    const formState = useFormState();
    const { validate: validateBillingAddressForm } = useFormApi();
    const { data } = useQuery(getCheckmoConfigQuery);

    const { data: isBillingAddressSameData } = useQuery(
        getIsBillingAddressSameQuery,
        { skip: !cartId, variables: { cartId } }
    );
    const { data: billingAddressData } = useQuery(getBillingAddressQuery, {
        skip: !cartId,
        variables: { cartId }
    });
    const { data: shippingAddressData } = useQuery(getShippingAddressQuery, {
        skip: !cartId,
        variables: { cartId }
    });
    const [
        updateBillingAddress,
        {
            error: billingAddressMutationError,
            called: billingAddressMutationCalled,
            loading: billingAddressMutationLoading
        }
    ] = useMutation(setBillingAddressMutation);
    const [
        setPaymentMethod,
        {
            error: paymentMethodMutationError,
            called: paymentMethodMutationCalled,
            loading: paymentMethodMutationLoading
        }
    ] = useMutation(setPaymentMethodOnCartMutation);

    const initialValues = useMemo(() => {
        const isBillingAddressSame = isBillingAddressSameData
            ? isBillingAddressSameData.cart.isBillingAddressSame
            : true;

        let billingAddress = {};
        /**
         * If billing address is same as shipping address, do
         * not auto fill the fields.
         */
        if (billingAddressData && !isBillingAddressSame) {
            if (billingAddressData.cart.billingAddress) {
                const {
                    // eslint-disable-next-line no-unused-vars
                    __typename,
                    ...rawBillingAddress
                } = billingAddressData.cart.billingAddress;
                billingAddress = mapAddressData(rawBillingAddress);
            }
        }

        return { isBillingAddressSame, ...billingAddress };
    }, [isBillingAddressSameData, billingAddressData]);

    const errors = useMemo(() => new Map([]), []);

    /*
     * Derived State.
     */
    const configuredMailingAddress =
        data &&
        data.storeConfig &&
        data.storeConfig.payment_checkmo_mailing_address;
    const mailingAddress = configuredMailingAddress
        ? configuredMailingAddress
        : propsMailingAddress;

    const shippingAddressCountry = shippingAddressData
        ? shippingAddressData.cart.shippingAddresses[0].country.code
        : DEFAULT_COUNTRY_CODE;

    const isBillingAddressSame = formState.values.isBillingAddressSame;

    /**
     * This function sets the boolean isBillingAddressSame
     * in cache for future use. We use cache because there
     * is no way to save this on the cart in remote.
     */
    const setIsBillingAddressSameInCache = useCallback(() => {
        client.writeQuery({
            query: getIsBillingAddressSameQuery,
            data: {
                cart: {
                    __typename: 'Cart',
                    id: cartId,
                    isBillingAddressSame
                }
            }
        });
    }, [client, cartId, getIsBillingAddressSameQuery, isBillingAddressSame]);

    /**
     * This function sets the billing address on the cart using the
     * information from the form.
     */
    const setBillingAddress = useCallback(() => {
        const {
            firstName,
            lastName,
            country,
            street1,
            street2,
            city,
            region,
            postcode,
            phoneNumber
        } = formState.values;

        updateBillingAddress({
            variables: {
                cartId,
                firstName,
                lastName,
                country,
                street1,
                street2,
                city,
                region: region.region_id,
                postcode,
                phoneNumber,
                sameAsShipping: false
            }
        });
    }, [formState.values, updateBillingAddress, cartId]);

    /**
     * This function sets the billing address on the cart using the
     * shipping address.
     */
    const setShippingAddressAsBillingAddress = useCallback(() => {
        const shippingAddress = shippingAddressData
            ? mapAddressData(shippingAddressData.cart.shippingAddresses[0])
            : {};

        updateBillingAddress({
            variables: {
                cartId,
                ...shippingAddress,
                sameAsShipping: true
            }
        });
    }, [updateBillingAddress, shippingAddressData, cartId]);

    /**
     * This function saves the payment method to the cart.
     */
    const setPaymentMethodOnCart = useCallback(() => {
        setPaymentMethod({ variables: { cartId } });
    }, [cartId, setPaymentMethod]);

    /*
     *  Effects.
     */

    useEffect(() => {
        if (!shouldSubmit) {
            console.log('Effect fired but we should not submit');
            return;
        }

        try {
            // Validate the form.
            validateBillingAddressForm();
            const hasErrors = Object.keys(formState.errors).length;
            if (hasErrors) {
                throw new Error('Errors in the billing address form');
            }

            // Set the billing address.
            setIsBillingAddressSameInCache();
            if (isBillingAddressSame) {
                setShippingAddressAsBillingAddress();
            } else {
                setBillingAddress();
            }

            // Set the payment method.
            setPaymentMethodOnCart();
        } catch (e) {
            if (process.env.NODE_ENV !== 'production') {
                console.error(err);
            }
            resetShouldSubmit();
        }
    }, [
        formState.errors,
        isBillingAddressSame,
        resetShouldSubmit,
        setBillingAddress,
        setIsBillingAddressSameInCache,
        setPaymentMethodOnCart,
        shouldSubmit,
        validateBillingAddressForm
    ]);

    useEffect(() => {
        const paymentMethodMutationCompleted =
            paymentMethodMutationCalled && !paymentMethodMutationLoading;

        if (paymentMethodMutationCompleted && !paymentMethodMutationError) {
            resetShouldSubmit();
            onPaymentSuccess();
        }

        if (paymentMethodMutationCompleted && paymentMethodMutationError) {
            resetShouldSubmit();
            onPaymentError();
        }
    }, [
        paymentMethodMutationError,
        paymentMethodMutationLoading,
        paymentMethodMutationCalled,
        onPaymentSuccess,
        onPaymentError,
        resetShouldSubmit
    ]);

    return {
        errors,
        initialValues,
        isBillingAddressSame,
        mailingAddress,
        payableTo:
            data &&
            data.storeConfig &&
            data.storeConfig.payment_checkmo_payable_to,
        shippingAddressCountry
    };
};
