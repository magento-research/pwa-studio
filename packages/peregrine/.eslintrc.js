const config = {
    parser: 'babel-eslint',
    extends: ['@magento'],
    rules: {
        "no-undef": "off"
    }
};

module.exports = config;
