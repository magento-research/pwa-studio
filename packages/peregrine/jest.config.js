module.exports = {
    displayName: 'Peregrine',
    setupFiles: ['<rootDir>/scripts/shim.js', '<rootDir>/scripts/fetch-mock.js']
};
