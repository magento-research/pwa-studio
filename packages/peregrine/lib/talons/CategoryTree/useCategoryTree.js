import { useEffect, useMemo } from 'react';
import { useQuery } from '../../hooks/useQuery';

export const useCategoryTree = props => {
    const { categories, categoryId, query, updateCategories } = props;

    const [queryResult, queryApi] = useQuery(query);
    const { data } = queryResult;
    const { runQuery } = queryApi;

    // fetch categories
    useEffect(() => {
        if (categoryId != null) {
            runQuery({ variables: { id: categoryId } });
        }
    }, [categoryId, runQuery]);

    // update redux with fetched categories
    useEffect(() => {
        if (data && data.category) {
            updateCategories(data.category);
        }
    }, [data, updateCategories]);

    const rootCategory = categories[categoryId];
    const { children } = rootCategory || {};

    const childCategories = useMemo(() => {
        const childCategories = new Map();

        for (const id of children || '') {
            const category = categories[id];
            const isLeaf = category.children_count === '0';

            childCategories.set(id, { category, isLeaf });
        }

        return childCategories;
    }, [categories, children]);

    return { childCategories };
};
