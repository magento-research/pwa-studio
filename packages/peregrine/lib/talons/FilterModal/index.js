export { useFilterBlock } from './useFilterBlock';
export { useFilterModal } from './useFilterModal';
export { useFilterState } from './useFilterState';
