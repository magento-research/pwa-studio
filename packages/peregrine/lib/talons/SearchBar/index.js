export { useAutocomplete } from './useAutocomplete';
export { useSearchBar } from './useSearchBar';
export { useSearchField } from './useSearchField';
export { useSuggestions } from './useSuggestions';
