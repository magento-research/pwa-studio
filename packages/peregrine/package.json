{
  "name": "@magento/peregrine",
  "version": "3.0.0",
  "publishConfig": {
    "access": "public"
  },
  "description": "The core runtime of Magento PWA",
  "main": "lib/index.js",
  "scripts": {
    "build": " ",
    "build:ci": " ",
    "build:cjs": " ",
    "build:dev": " ",
    "clean": " ",
    "storybook": "start-storybook -p 9001 -c .storybook",
    "storybook:build": "build-storybook -c .storybook -o storybook-dist",
    "watch": " "
  },
  "repository": "github:magento-research/pwa-studio",
  "author": "Magento Commerce",
  "license": "(OSL-3.0 OR AFL-3.0)",
  "bugs": {
    "url": "https://github.com/magento-research/pwa-studio/issues"
  },
  "homepage": "https://github.com/magento-research/pwa-studio/tree/master/packages/peregrine#readme",
  "devDependencies": {
    "@storybook/react": "~4.1.13",
    "intl": "~1.2.5",
    "react": "~16.8.3",
    "react-apollo": "~2.5.3",
    "react-dom": "~16.8.3",
    "react-router-dom": "~5.0.0",
    "react-test-renderer": "~16.8.6",
    "rimraf": "~2.6.3"
  },
  "peerDependencies": {
    "@babel/runtime": "^7.4.2",
    "react": "^16.8.0",
    "react-apollo": "^2.5.3",
    "react-dom": "^16.8.0",
    "react-router-dom": "~5.0.0"
  },
  "engines": {
    "node": "8.10.0 || >=10.7.0",
    "yarn": ">=1.12.0"
  },
  "module": "lib/index.js",
  "jsnext:main": "lib/index.js",
  "es2015": "lib/index.js",
  "sideEffects": false
}
