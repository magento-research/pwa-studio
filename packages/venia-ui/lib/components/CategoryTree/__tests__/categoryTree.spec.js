import React from 'react';
import { createTestInstance } from '@magento/peregrine';
import { useCategoryTree } from '@magento/peregrine/lib/talons/CategoryTree';

import CategoryTree from '../categoryTree';

jest.mock('../../../classify');
jest.mock('../categoryBranch', () => () => <i />);
jest.mock('../categoryLeaf', () => () => <i />);

jest.mock('@magento/peregrine/lib/talons/CategoryTree', () => {
    const childCategories = new Map();
    const useCategoryTree = jest.fn(() => ({ childCategories }));

    return { useCategoryTree };
});

const props = {
    categories: {},
    categoryId: 1,
    onNavigate: jest.fn(),
    setCategoryId: jest.fn(),
    updateCategories: jest.fn()
};

test('renders correctly without data', async () => {
    const instance = createTestInstance(<CategoryTree {...props} />);

    expect(instance.toJSON()).toMatchSnapshot();
});

test('calls talon with correct props', async () => {
    createTestInstance(<CategoryTree {...props} />);

    expect(useCategoryTree).toHaveBeenCalledTimes(1);
    expect(useCategoryTree).toHaveBeenNthCalledWith(1, {
        categories: props.categories,
        categoryId: props.categoryId,
        query: expect.any(Object),
        updateCategories: props.updateCategories
    });
});

test('calls childCategories from talon props', async () => {
    useCategoryTree.mockReturnValue({
        childCategories: new Map([
            ['1', { category: 'Shirts', isLeaf: false, parentCategory: null }],
            ['2', { category: 'Short Sleeve', isLeaf: true, parentCategory: 1 }]
        ])
    });

    const instance = createTestInstance(<CategoryTree {...props} />);
    expect(instance.toJSON()).toMatchSnapshot();
});
