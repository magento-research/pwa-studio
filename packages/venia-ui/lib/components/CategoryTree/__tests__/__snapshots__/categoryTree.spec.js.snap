// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calls childCategories from talon props 1`] = `
<div
  className="root"
>
  <ul
    className="tree"
  >
    <i />
    <i />
    <i />
  </ul>
</div>
`;

exports[`renders correctly without data 1`] = `
<div
  className="root"
>
  <ul
    className="tree"
  />
</div>
`;
